Flaky-test: org.apache.pulsar.broker.service.ConsumedLedgersTrimTest.testConsumedLedgersTrimNoSubscriptions

org.apache.pulsar.broker.service.ConsumedLedgersTrimTest is flaky. The testConsumedLedgersTrimNoSubscriptions test method fails sporadically.

```
org.apache.pulsar.client.admin.PulsarAdminException$ServerSideErrorException: HTTP 500 Internal Server Error
	at org.apache.pulsar.client.admin.internal.BaseResource.getApiException(BaseResource.java:205)
	at org.apache.pulsar.client.admin.internal.TopicsImpl$10.failed(TopicsImpl.java:657)
	at org.glassfish.jersey.client.JerseyInvocation$1.failed(JerseyInvocation.java:839)
	at org.glassfish.jersey.client.JerseyInvocation$1.completed(JerseyInvocation.java:820)
	at org.glassfish.jersey.client.ClientRuntime.processResponse(ClientRuntime.java:229)
	at org.glassfish.jersey.client.ClientRuntime.access$200(ClientRuntime.java:62)
	at org.glassfish.jersey.client.ClientRuntime$2.lambda$response$0(ClientRuntime.java:173)
	at org.glassfish.jersey.internal.Errors$1.call(Errors.java:248)
	at org.glassfish.jersey.internal.Errors$1.call(Errors.java:244)
	at org.glassfish.jersey.internal.Errors.process(Errors.java:292)
	at org.glassfish.jersey.internal.Errors.process(Errors.java:274)
	at org.glassfish.jersey.internal.Errors.process(Errors.java:244)
	at org.glassfish.jersey.process.internal.RequestScope.runInScope(RequestScope.java:288)
	at org.glassfish.jersey.client.ClientRuntime$2.response(ClientRuntime.java:173)
```
Usage tip: To enable automatic navigation to failure message, open the links with CTRL/CMD-click.
[example failure 2021-02-01T13:27:38.6355573Z](https://github.com/apache/pulsar/runs/1806657480?check_suite_focus=true#step:9:1303)
[example failure 2021-02-01T13:24:05.4419686Z](https://github.com/apache/pulsar/runs/1806383882?check_suite_focus=true#step:9:1352)
[example failure 2021-02-01T13:18:55.4690491Z](https://github.com/apache/pulsar/runs/1806657480?check_suite_focus=true#step:9:759)
[example failure 2021-02-01T13:14:49.3799980Z](https://github.com/apache/pulsar/runs/1806383882?check_suite_focus=true#step:9:763)
[example failure 2021-02-01T11:41:54.2894143Z](https://github.com/apache/pulsar/runs/1805502931?check_suite_focus=true#step:9:1332)
[example failure 2021-02-01T11:33:02.6744190Z](https://github.com/apache/pulsar/runs/1805502931?check_suite_focus=true#step:9:757)
[example failure 2021-02-01T11:10:53.1246465Z](https://github.com/apache/pulsar/runs/1805264316?check_suite_focus=true#step:9:1621)
[example failure 2021-02-01T11:02:12.4881454Z](https://github.com/apache/pulsar/runs/1805264316?check_suite_focus=true#step:9:1079)
[example failure 2021-02-01T10:54:17.7577015Z](https://github.com/apache/pulsar/runs/1805085533?check_suite_focus=true#step:9:1487)
[example failure 2021-02-01T10:45:31.1222974Z](https://github.com/apache/pulsar/runs/1805085533?check_suite_focus=true#step:9:916)
[example failure 2021-02-01T04:03:27.6805735Z](https://github.com/apache/pulsar/runs/1803720627?check_suite_focus=true#step:9:2013)
[example failure 2021-02-01T03:53:05.2559217Z](https://github.com/apache/pulsar/runs/1803720627?check_suite_focus=true#step:9:1450)
[example failure 2021-02-01T03:43:03.9441607Z](https://github.com/apache/pulsar/runs/1803720627?check_suite_focus=true#step:9:904)
[example failure 2021-02-01T03:30:07.1668311Z](https://github.com/apache/pulsar/runs/1803586038?check_suite_focus=true#step:9:2714)
[example failure 2021-01-30T18:42:32.1891764Z](https://github.com/apache/pulsar/runs/1798267292?check_suite_focus=true#step:9:2290)
[example failure 2021-01-30T18:26:20.0175309Z](https://github.com/apache/pulsar/runs/1798267292?check_suite_focus=true#step:9:1230)
[example failure 2021-01-30T09:58:38.1619061Z](https://github.com/apache/pulsar/runs/1796827745?check_suite_focus=true#step:9:2685)
[example failure 2021-01-30T09:49:25.7784734Z](https://github.com/apache/pulsar/runs/1796827745?check_suite_focus=true#step:9:1904)
[example failure 2021-01-30T09:07:26.4235087Z](https://github.com/apache/pulsar/runs/1796714099?check_suite_focus=true#step:9:2800)
[example failure 2021-01-30T08:58:16.1045896Z](https://github.com/apache/pulsar/runs/1796714099?check_suite_focus=true#step:9:1854)
[example failure 2021-01-29T00:24:21.7394038Z](https://github.com/apache/pulsar/runs/1788100574?check_suite_focus=true#step:9:2045)
[example failure 2021-01-29T00:16:30.7401540Z](https://github.com/apache/pulsar/runs/1788100574?check_suite_focus=true#step:9:1480)
[example failure 2021-01-29T00:08:32.0451653Z](https://github.com/apache/pulsar/runs/1788100574?check_suite_focus=true#step:9:916)
[example failure 2021-01-28T18:34:02.9858608Z](https://github.com/apache/pulsar/runs/1784535146?check_suite_focus=true#step:8:1761)
[example failure 2021-01-28T07:24:30.1532293Z](https://github.com/apache/pulsar/runs/1782001667?check_suite_focus=true#step:8:2004)
[example failure 2021-01-28T07:16:39.1861158Z](https://github.com/apache/pulsar/runs/1782001667?check_suite_focus=true#step:8:1436)
[example failure 2021-01-28T01:53:06.1986225Z](https://github.com/apache/pulsar/runs/1781059882?check_suite_focus=true#step:8:1048)
[example failure 2021-01-27T23:24:00.0650893Z](https://github.com/apache/pulsar/runs/1779704427?check_suite_focus=true#step:8:1463)
[example failure 2021-01-27T22:07:27.7334507Z](https://github.com/apache/pulsar/runs/1778467744?check_suite_focus=true#step:8:1996)
[example failure 2021-01-27T21:51:26.4006342Z](https://github.com/apache/pulsar/runs/1778467744?check_suite_focus=true#step:8:931)
[example failure 2021-01-27T20:55:51.0401183Z](https://github.com/apache/pulsar/runs/1778222671?check_suite_focus=true#step:8:2396)
[example failure 2021-01-27T20:40:37.9711910Z](https://github.com/apache/pulsar/runs/1778222671?check_suite_focus=true#step:8:1348)
[example failure 2021-01-27T13:13:57.5683990Z](https://github.com/apache/pulsar/runs/1775226632?check_suite_focus=true#step:8:1750)
[example failure 2021-01-27T13:06:25.9941350Z](https://github.com/apache/pulsar/runs/1775226632?check_suite_focus=true#step:8:1200)
[example failure 2021-01-27T11:59:57.1093570Z](https://github.com/apache/pulsar/runs/1775096684?check_suite_focus=true#step:8:1451)
[example failure 2021-01-27T11:52:18.6018485Z](https://github.com/apache/pulsar/runs/1775096684?check_suite_focus=true#step:8:889)
[example failure 2021-01-27T06:40:57.0530564Z](https://github.com/apache/pulsar/runs/1774699381?check_suite_focus=true#step:8:1466)
[example failure 2021-01-27T01:34:33.3416104Z](https://github.com/apache/pulsar/runs/1773579903?check_suite_focus=true#step:8:1336)
[example failure 2021-01-27T01:26:54.3459256Z](https://github.com/apache/pulsar/runs/1773579903?check_suite_focus=true#step:8:768)


<details>
<summary>Full exception stacktrace</summary>
<code><pre>
org.apache.pulsar.client.admin.PulsarAdminException$ServerSideErrorException: HTTP 500 Internal Server Error
	at org.apache.pulsar.client.admin.internal.BaseResource.getApiException(BaseResource.java:205)
	at org.apache.pulsar.client.admin.internal.TopicsImpl$10.failed(TopicsImpl.java:657)
	at org.glassfish.jersey.client.JerseyInvocation$1.failed(JerseyInvocation.java:839)
	at org.glassfish.jersey.client.JerseyInvocation$1.completed(JerseyInvocation.java:820)
	at org.glassfish.jersey.client.ClientRuntime.processResponse(ClientRuntime.java:229)
	at org.glassfish.jersey.client.ClientRuntime.access$200(ClientRuntime.java:62)
	at org.glassfish.jersey.client.ClientRuntime$2.lambda$response$0(ClientRuntime.java:173)
	at org.glassfish.jersey.internal.Errors$1.call(Errors.java:248)
	at org.glassfish.jersey.internal.Errors$1.call(Errors.java:244)
	at org.glassfish.jersey.internal.Errors.process(Errors.java:292)
	at org.glassfish.jersey.internal.Errors.process(Errors.java:274)
	at org.glassfish.jersey.internal.Errors.process(Errors.java:244)
	at org.glassfish.jersey.process.internal.RequestScope.runInScope(RequestScope.java:288)
	at org.glassfish.jersey.client.ClientRuntime$2.response(ClientRuntime.java:173)
	at org.apache.pulsar.client.admin.internal.http.AsyncHttpConnector.lambda$apply$1(AsyncHttpConnector.java:210)
	at java.util.concurrent.CompletableFuture.uniWhenComplete(CompletableFuture.java:774)
	at java.util.concurrent.CompletableFuture$UniWhenComplete.tryFire(CompletableFuture.java:750)
	at java.util.concurrent.CompletableFuture.postComplete(CompletableFuture.java:488)
	at java.util.concurrent.CompletableFuture.complete(CompletableFuture.java:1975)
	at org.apache.pulsar.client.admin.internal.http.AsyncHttpConnector.lambda$retryOperation$3(AsyncHttpConnector.java:251)
	at java.util.concurrent.CompletableFuture.uniWhenComplete(CompletableFuture.java:774)
	at java.util.concurrent.CompletableFuture$UniWhenComplete.tryFire(CompletableFuture.java:750)
	at java.util.concurrent.CompletableFuture.postComplete(CompletableFuture.java:488)
	at java.util.concurrent.CompletableFuture.complete(CompletableFuture.java:1975)
	at org.asynchttpclient.netty.NettyResponseFuture.loadContent(NettyResponseFuture.java:222)
	at org.asynchttpclient.netty.NettyResponseFuture.done(NettyResponseFuture.java:257)
	at org.asynchttpclient.netty.handler.AsyncHttpClientHandler.finishUpdate(AsyncHttpClientHandler.java:241)
	at org.asynchttpclient.netty.handler.HttpHandler.handleChunk(HttpHandler.java:114)
	at org.asynchttpclient.netty.handler.HttpHandler.handleRead(HttpHandler.java:143)
	at org.asynchttpclient.netty.handler.AsyncHttpClientHandler.channelRead(AsyncHttpClientHandler.java:78)
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:379)
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:365)
	at io.netty.channel.AbstractChannelHandlerContext.fireChannelRead(AbstractChannelHandlerContext.java:357)
	at io.netty.handler.codec.MessageToMessageDecoder.channelRead(MessageToMessageDecoder.java:103)
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:379)
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:365)
	at io.netty.channel.AbstractChannelHandlerContext.fireChannelRead(AbstractChannelHandlerContext.java:357)
	at io.netty.channel.CombinedChannelDuplexHandler$DelegatingChannelHandlerContext.fireChannelRead(CombinedChannelDuplexHandler.java:436)
	at io.netty.handler.codec.ByteToMessageDecoder.fireChannelRead(ByteToMessageDecoder.java:324)
	at io.netty.handler.codec.ByteToMessageDecoder.channelRead(ByteToMessageDecoder.java:296)
	at io.netty.channel.CombinedChannelDuplexHandler.channelRead(CombinedChannelDuplexHandler.java:251)
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:379)
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:365)
	at io.netty.channel.AbstractChannelHandlerContext.fireChannelRead(AbstractChannelHandlerContext.java:357)
	at io.netty.channel.DefaultChannelPipeline$HeadContext.channelRead(DefaultChannelPipeline.java:1410)
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:379)
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:365)
	at io.netty.channel.DefaultChannelPipeline.fireChannelRead(DefaultChannelPipeline.java:919)
	at io.netty.channel.nio.AbstractNioByteChannel$NioByteUnsafe.read(AbstractNioByteChannel.java:163)
	at io.netty.channel.nio.NioEventLoop.processSelectedKey(NioEventLoop.java:714)
	at io.netty.channel.nio.NioEventLoop.processSelectedKeysOptimized(NioEventLoop.java:650)
	at io.netty.channel.nio.NioEventLoop.processSelectedKeys(NioEventLoop.java:576)
	at io.netty.channel.nio.NioEventLoop.run(NioEventLoop.java:493)
	at io.netty.util.concurrent.SingleThreadEventExecutor$4.run(SingleThreadEventExecutor.java:989)
	at io.netty.util.internal.ThreadExecutorMap$2.run(ThreadExecutorMap.java:74)
	at io.netty.util.concurrent.FastThreadLocalRunnable.run(FastThreadLocalRunnable.java:30)
	at java.lang.Thread.run(Thread.java:748)
Caused by: javax.ws.rs.InternalServerErrorException: HTTP 500 Internal Server Error
	at org.glassfish.jersey.client.JerseyInvocation.convertToException(JerseyInvocation.java:914)
	at org.glassfish.jersey.client.JerseyInvocation.access$500(JerseyInvocation.java:77)
	... 54 more

</pre></code>
</details>

```
java.lang.AssertionError: expected [3:6:-1:0] but found [-1:-1:-1]
	at org.testng.Assert.fail(Assert.java:99)
	at org.testng.Assert.failNotEquals(Assert.java:1037)
	at org.testng.Assert.assertEqualsImpl(Assert.java:140)
	at org.testng.Assert.assertEquals(Assert.java:122)
	at org.testng.Assert.assertEquals(Assert.java:617)
	at org.apache.pulsar.broker.service.ConsumedLedgersTrimTest.testConsumedLedgersTrimNoSubscriptions(ConsumedLedgersTrimTest.java:148)
```

[example failure 2021-01-28T15:19:05.1394802Z](https://github.com/apache/pulsar/runs/1784405661?check_suite_focus=true#step:8:1251)


<details>
<summary>Full exception stacktrace</summary>
<code><pre>
java.lang.AssertionError: expected [3:6:-1:0] but found [-1:-1:-1]
	at org.testng.Assert.fail(Assert.java:99)
	at org.testng.Assert.failNotEquals(Assert.java:1037)
	at org.testng.Assert.assertEqualsImpl(Assert.java:140)
	at org.testng.Assert.assertEquals(Assert.java:122)
	at org.testng.Assert.assertEquals(Assert.java:617)
	at org.apache.pulsar.broker.service.ConsumedLedgersTrimTest.testConsumedLedgersTrimNoSubscriptions(ConsumedLedgersTrimTest.java:148)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:498)
	at org.testng.internal.MethodInvocationHelper.invokeMethod(MethodInvocationHelper.java:132)
	at org.testng.internal.InvokeMethodRunnable.runOne(InvokeMethodRunnable.java:45)
	at org.testng.internal.InvokeMethodRunnable.call(InvokeMethodRunnable.java:73)
	at org.testng.internal.InvokeMethodRunnable.call(InvokeMethodRunnable.java:11)
	at java.util.concurrent.FutureTask.run(FutureTask.java:266)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
	at java.lang.Thread.run(Thread.java:748)

</pre></code>
</details>

--- a/pulsar-broker/src/test/java/org/apache/pulsar/broker/service/ConsumedLedgersTrimTest.java
+++ b/pulsar-broker/src/test/java/org/apache/pulsar/broker/service/ConsumedLedgersTrimTest.java
@@ -31,6 +31,7 @@ import org.apache.pulsar.client.api.Consumer;
 import org.apache.pulsar.client.api.Message;
 import org.apache.pulsar.client.api.MessageId;
 import org.apache.pulsar.client.api.Producer;
+import org.awaitility.Awaitility;
 import org.slf4j.Logger;
 import org.slf4j.LoggerFactory;
 import org.testng.Assert;
@@ -142,6 +143,14 @@ public class ConsumedLedgersTrimTest extends BrokerTestBase {
         // the lastMessageId is still on the previous ledger
         restartBroker();
         // force load topic
+        Awaitility.await().atMost(5, TimeUnit.SECONDS).until(() -> {
+            try {
+                pulsar.getBrokerService().getTopicIfExists(topicName).get(3, TimeUnit.SECONDS).get();
+                return true;
+            } catch (Exception ignored) {
+            }
+            return false;
+        });
         pulsar.getAdminClient().topics().getStats(topicName);
         MessageId messageIdAfterRestart = pulsar.getAdminClient().topics().getLastMessageId(topicName);
         LOG.info("lastmessageid " + messageIdAfterRestart);

--- a/pulsar-broker/src/test/java/org/apache/pulsar/broker/service/ConsumedLedgersTrimTest.java
+++ b/pulsar-broker/src/test/java/org/apache/pulsar/broker/service/ConsumedLedgersTrimTest.java
@@ -21,6 +21,8 @@ package org.apache.pulsar.broker.service;
 
 import static org.testng.Assert.assertEquals;
 import static org.testng.Assert.assertNotEquals;
+import static org.testng.Assert.assertNotNull;
+
 import java.util.concurrent.CompletableFuture;
 import java.util.concurrent.TimeUnit;
 import lombok.Cleanup;
@@ -143,14 +145,8 @@ public class ConsumedLedgersTrimTest extends BrokerTestBase {
         // the lastMessageId is still on the previous ledger
         restartBroker();
         // force load topic
-        Awaitility.await().atMost(5, TimeUnit.SECONDS).until(() -> {
-            try {
-                pulsar.getBrokerService().getTopicIfExists(topicName).get(3, TimeUnit.SECONDS).get();
-                return true;
-            } catch (Exception ignored) {
-            }
-            return false;
-        });
+        Awaitility.await().ignoreExceptions().untilAsserted(()
+                -> assertNotNull(pulsar.getBrokerService().getTopicIfExists(topicName).get(3, TimeUnit.SECONDS).get()));
         pulsar.getAdminClient().topics().getStats(topicName);
         MessageId messageIdAfterRestart = pulsar.getAdminClient().topics().getLastMessageId(topicName);
         LOG.info("lastmessageid " + messageIdAfterRestart);

